### ISSUE TITLE
kube-vip requires super-admin.conf with Kubernetes 1.29

### ISSUE THREAD

**Issue by louhisuo** [kube-vip/kube-vip#684]
**Describe the bug**
Setting up a new Kubernetes 1.29.0 cluster with kubeadm fails when kube-vip is used. Lima-VM with Ubuntu 22.04 LTS image was used as a machine.

**To Reproduce**
1. Start a machine (virtual machine) with all Kubernetes prerequisites satisfied for Kubernetes 1.29.0

2. Within a control plane machine generate kube-vip static pod
```
export KVVERSION=v0.6.4
export INTERFACE=lima0
export VIP=<VIP address>
sudo ctr image pull ghcr.io/kube-vip/kube-vip:$KVVERSION
sudo ctr run --rm --net-host ghcr.io/kube-vip/kube-vip:$KVVERSION vip /kube-vip manifest pod \
    --arp \
    --controlplane \
    --address $VIP \
    --interface $INTERFACE \
    --enableLoadBalancer \
    --leaderElection | sudo tee /etc/kubernetes/manifests/kube-vip.yaml

```
4. Initiate Kubernetes 1.29.0 control plane with 'kubeadm init', observe timeout and failure to initiate a cluster
5. Check logs for kube-vip container with crictl and observe an error
```
# crictl --runtime-endpoint unix:///var/run/containerd/containerd.sock logs 72f0058318b86
time="2023-12-14T09:30:38Z" level=info msg="Starting kube-vip.io [v0.6.4]"
time="2023-12-14T09:30:38Z" level=info msg="namespace [kube-system], Mode: [ARP], Features(s): Control Plane:[true], Services:[false]"
time="2023-12-14T09:30:38Z" level=info msg="prometheus HTTP server started"
time="2023-12-14T09:30:38Z" level=info msg="Starting Kube-vip Manager with the ARP engine"
time="2023-12-14T09:30:38Z" level=info msg="Beginning cluster membership, namespace [kube-system], lock name [plndr-cp-lock], id [lima-cp-1]"
I1214 09:30:38.362358       1 leaderelection.go:250] attempting to acquire leader lease kube-system/plndr-cp-lock...
E1214 09:30:38.907904       1 leaderelection.go:332] error retrieving resource lock kube-system/plndr-cp-lock: leases.coordination.k8s.io "plndr-cp-lock" is forbidden: User "kubernetes-admin" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kube-system"
E1214 09:30:40.258727       1 leaderelection.go:332] error retrieving resource lock kube-system/plndr-cp-lock: leases.coordination.k8s.io "plndr-cp-lock" is forbidden: User "kubernetes-admin" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kube-system"
```

**Expected behavior**
kube-vip is able to retrieve resource lock. This can be achieved by manually editing kube-vip.conf with following configuration before running `kubeadm init`.
```
  volumes:
  - hostPath:
      path: /etc/kubernetes/super-admin.conf
    name: kubeconfig
```
For details why see [Kubernetes 1.29 CHANGELOG](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.29.md#no-really-you-must-read-this-before-you-upgrade)

**Screenshots**
n/a

**Environment (please complete the following information):**
 - OS/Distro: Ubuntu 22.04 LTS
 - Kubernetes Version: 1.29.0
 - Kube-vip Version: 0.6.4

**`Kube-vip.yaml`:**
Will be provided if really needed.

**Additional context**
kubeadm init generates another kubeconfig file super-admin.conf in Kubernetes 1.29 and onwards which is required by kube-vip static pod to able start static pod during cluster initiation.

**Comment by tuxtof** [comment]
problem is that super-admin.conf is only present on the first control plane

**Comment by louhisuo** [comment]
Not sure if it helps you forward but there is kubeadm issue [#2414](https://github.com/kubernetes/kubeadm/issues/2414) which seems to describe this feature.

**Comment by neolit123** [comment]
this sounds like a bootstrap problem?

1.29 kubeadm init is doing the following:
- writes an admin.conf that has no binding to cluster-admin yet
- writes a super-admin.conf that has system:masters (super user)
- create cluster-admin binding using the super-admin

super-admin is not meant to be used by tools like kube-vip.
if kube-vip needs permissiosn during setup it can, either:
- wait for admin.conf to receive permissions (not possible AFAIK, given the bootstrap sequence)
- use super-admin.conf during bootstrap and then moving to admin.conf

**Comment by neolit123** [comment]
let's see what is actionable and we may have to also update these kubeadm docs:
https://github.com/kubernetes/kubeadm/blob/main/docs/ha-considerations.md#kube-vip

**Comment by chrischdi** [comment]
Workaround:
* use `super-admin.conf` during `kubeadm init`.
* revert the static pod back to `admin.conf` after successful `kubeadm init` (which causes a recreation of the kube-vip pod)

Example solution:
* command pre-kubeadm:
  ```
  sed -i 's#path: /etc/kubernetes/admin.conf#path: /etc/kubernetes/super-admin.conf#' \
            /etc/kubernetes/manifests/kube-vip.yaml
  ```
* command post-kubeadm:
  ```
  sed -i 's#path: /etc/kubernetes/super-admin.conf#path: /etc/kubernetes/admin.conf#' \
            /etc/kubernetes/manifests/kube-vip.yaml
  ```

**Comment by prezha** [comment]
> Workaround:
> 
> * use `super-admin.conf` during `kubeadm init`.
> * revert the static pod back to `admin.conf` after successful `kubeadm init` (which causes a recreation of the kube-vip pod)
> 
> Example solution (only run on the node which runs `kubeadm init`):
> 
> * command pre-kubeadm:
>   ```
>   sed -i 's#path: /etc/kubernetes/admin.conf#path: /etc/kubernetes/super-admin.conf#' \
>             /etc/kubernetes/manifests/kube-vip.yaml
>   ```
> * command post-kubeadm:
>   ```
>   sed -i 's#path: /etc/kubernetes/super-admin.conf#path: /etc/kubernetes/admin.conf#' \
>             /etc/kubernetes/manifests/kube-vip.yaml
>   ```

@chrischdi the above override should only apply to the `hostPath` and not the `mountPath` (that should stay `/etc/kubernetes/admin.conf` in either case)?

https://github.com/kube-vip/kube-vip/blob/3c0e5e327e21e26f1f131c90e0908dc01662ae37/cmd/kube-vip.go#L153

**Comment by Jeansen** [comment]
Does not seem to be resolved. Creating a plain Kubernetes cluster with 1.28 works like a charm. Using Kubenetes 1.29 still causes problems during kubeadm init. What's the status on this? I am a little bit puzzled, because I see there has been some work merged into the latest releases. But still, even in v0.7.2, I cannot create a cluster with static kube-vip pods.

**Comment by Jeansen** [comment]
> Workaround:
> 
> * use `super-admin.conf` during `kubeadm init`.
> * revert the static pod back to `admin.conf` after successful `kubeadm init` (which causes a recreation of the kube-vip pod)
> 
> Example solution (only run on the node which runs `kubeadm init`):
> 
> * command pre-kubeadm:
>   ```
>   sed -i 's#path: /etc/kubernetes/admin.conf#path: /etc/kubernetes/super-admin.conf#' \
>             /etc/kubernetes/manifests/kube-vip.yaml
>   ```
> * command post-kubeadm (Edit note: this causes a pod restart and may cause flaky behavior):
>   ```
>   sed -i 's#path: /etc/kubernetes/super-admin.conf#path: /etc/kubernetes/admin.conf#' \
>             /etc/kubernetes/manifests/kube-vip.yaml
>   ```

That worked, so far!

**Comment by limylily** [comment]
@Jeansen What would happen if I don't switch from `super-admin.conf`  to `admin. conf`  after `kubeadm init`

**Comment by Vigilans** [comment]
FYI: For a three-control-plane master, with @chrischdi's pre- and post- setup:
1. with 1 CP is down, the VIP can float
2. then, with 2 CP down kube-vip starts to complain because of etcd issue
3. finally, with 1 CP recovered and 1 CP down, the kube-vip once again starts to complain `error retrieving resource lock` like this issue describes. 

With super-admin.conf distributed and used in each CP's kube-vip pod, step 3 can works without issue. (I have tested these two setups only once, so the result may not be very reliable.)

**Comment by jjschwarz** [comment]
Any official method of handling this beyond the scripts above?

**Comment by chrischdi** [comment]
Note: with Kubernetes v1.31, the `ControlPlaneKubeletLocalMode` feature-gate for kubeadm could solve this issue so the workaround may not be required.

**Comment by neolit123** [comment]
> Note: with Kubernetes v1.31, the `ControlPlaneKubeletLocalMode` feature-gate for kubeadm could solve this issue so the workaround may not be required.

could you elaborate @chrischdi ?

**Comment by neolit123** [comment]
that makes sense, thanks for the explanation!

**Comment by chrischdi** [comment]
> Without `ControlPlaneKubeletLocalMode` and when referncing `admin.conf` for kube-vip:
> 
> * The kubelet got started
> * The kubelet wanted to bootstrap itself by using the control-plane IP
> * This failed until kube-vip comes up
> * The kubelet can't start kube-vip because the `admin.conf` does not yet exist
> 
> With `ControlPlaneKubeletLocalMode` and when referncing `admin.conf` for kube-vip:
> 
> * The kubelet got started
> * The kubelet bootstraps itself using the local control-plane IP (not depending on kube-vip being up)
> * The admin.conf gets created
> * The kubelet should be able to start kube-vip now

Looks like this does not work. Kubeadm needs to have the loadbalancer IP to be reachable before applying the rbac for the admin.conf:

cloud-init:
```
[2024-08-26 09:28:49] [certs] Using the existing "sa" key
[2024-08-26 09:28:49] [kubeconfig] Using kubeconfig folder "/etc/kubernetes"
[2024-08-26 09:28:49] [kubeconfig] Writing "admin.conf" kubeconfig file
[2024-08-26 09:28:49] [kubeconfig] Writing "super-admin.conf" kubeconfig file
[2024-08-26 09:28:49] [kubeconfig] Writing "kubelet.conf" kubeconfig file
[2024-08-26 09:28:50] [kubeconfig] Writing "controller-manager.conf" kubeconfig file
[2024-08-26 09:28:50] [kubeconfig] Writing "scheduler.conf" kubeconfig file
[2024-08-26 09:28:50] [etcd] Creating static Pod manifest for local etcd in "/etc/kubernetes/manifests"
[2024-08-26 09:28:50] [control-plane] Using manifest folder "/etc/kubernetes/manifests"
[2024-08-26 09:28:50] [control-plane] Creating static Pod manifest for "kube-apiserver"
[2024-08-26 09:28:50] [control-plane] Creating static Pod manifest for "kube-controller-manager"
[2024-08-26 09:28:50] [control-plane] Creating static Pod manifest for "kube-scheduler"
[2024-08-26 09:28:50] [kubelet-start] Writing kubelet environment file with flags to file "/var/lib/kubelet/kubeadm-flags.env"
[2024-08-26 09:28:50] [kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[2024-08-26 09:28:50] [kubelet-start] Starting the kubelet
[2024-08-26 09:28:50] [wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests"
[2024-08-26 09:28:50] [kubelet-check] Waiting for a healthy kubelet at http://127.0.0.1:10248/healthz. This can take up to 4m0s
[2024-08-26 09:28:51] [kubelet-check] The kubelet is healthy after 1.001231216s
[2024-08-26 09:28:51] [api-check] Waiting for a healthy API server. This can take up to 4m0s
[2024-08-26 09:32:51] [api-check] The API server is not healthy after 4m0.000302569s
[2024-08-26 09:32:51]
[2024-08-26 09:32:51] Unfortunately, an error has occurred:
[2024-08-26 09:32:51]   context deadline exceeded
[2024-08-26 09:32:51]
[2024-08-26 09:32:51] This error is likely caused by:
[2024-08-26 09:32:51]   - The kubelet is not running
[2024-08-26 09:32:51]   - The kubelet is unhealthy due to a misconfiguration of the node in some way (required cgroups disabled)
[2024-08-26 09:32:51]
[2024-08-26 09:32:51] If you are on a systemd-powered system, you can try to troubleshoot the error with the following commands:
[2024-08-26 09:32:51]   - 'systemctl status kubelet'
[2024-08-26 09:32:51]   - 'journalctl -xeu kubelet'
[2024-08-26 09:32:51]
[2024-08-26 09:32:51] Additionally, a control plane component may have crashed or exited when started by the container runtime.
[2024-08-26 09:32:51] To troubleshoot, list all containers using your preferred container runtimes CLI.
[2024-08-26 09:32:51] Here is one example how you may list all running Kubernetes containers by using crictl:
[2024-08-26 09:32:51]   - 'crictl --runtime-endpoint unix:///var/run/containerd/containerd.sock ps -a | grep kube | grep -v pause'
[2024-08-26 09:32:51]   Once you have found the failing container, you can inspect its logs with:
[2024-08-26 09:32:51]   - 'crictl --runtime-endpoint unix:///var/run/containerd/containerd.sock logs CONTAINERID'
[2024-08-26 09:32:51] error execution phase wait-control-plane: could not initialize a Kubernetes cluster
```

kube-vip:
```
E0826 09:35:44.260808       1 leaderelection.go:332] error retrieving resource lock kube-system/plndr-cp-lock: leases.coordination.k8s.io "plndr-cp-lock" is forbidden: User "kubernetes-admin" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kube-system"
```

**Comment by W1zzardTPU** [comment]
Could something like that help for bootstrap?

```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: kube-system
  name: kube-vip-lease-access-role
rules:
- apiGroups: ["coordination.k8s.io"]
  resources: ["leases"]
  verbs: ["get", "watch", "list", "create", "update", "patch", "delete"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: kube-vip-lease-access-binding
  namespace: kube-system
subjects:
- kind: ServiceAccount
  name: default
  namespace: kube-system
roleRef:
  kind: Role
  name: kube-vip-lease-access-role
  apiGroup: rbac.authorization.k8s.io
```

**Comment by chrischdi** [comment]
That's a chicken-egg issue. Who applies that?
The admin.conf has enough permissions, but kubeadm applies the rbac for that at a later stage and needs the control-plane IP connectivity first.

**Comment by rikatz** [comment]
didn't tested yet, but maybe instead of using superadmin, what would happen if we make kube-vip at first not rely on leader election (during init, but not during join) so it can start, then allow it to init, and post init we set leader election back?

IIRC init will run just during cluster bootstrap, which I guess will give kubeadm a chance to have a vip, and then add the right permissions to admin, and during join all of the other nodes will be already using the right permissions?

I will try to test it tomorrow or next week :)

**Comment by rikatz** [comment]
followup on my test, doesn't work as kube-vip still keeps to try to use leader election :(

**Comment by lubronzhan** [comment]
Can kubeadm talk to apiserver through localhost? Does kubeadm init always run on the control plane VM?

**Comment by thebsdbox** [comment]
I added some functionality for kube-vip to find a way to speak to the api-server. I never really finished it off though, let me find it.

**Comment by neolit123** [comment]
both the `admin.conf` and `superadmin.conf` talk to the control-plane-endpoint (LB/VIP):
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go#L490-L507

during `init` the admin/super-admin kubeconfigs could point to the local API server and are later mutated in a finalization phase to point to the LB. ideas/experimentation and PRs are welcome.

e.g. pseudo:
```
kubeadm init phase certs ca --config some-kubeadm-config.yaml
kubeadm init phase kubeconfig admin --config some-kubeadm-config.yaml
kubeadm init phase kubeconfig super-admin --config some-kubeadm-config.yaml
# sed the admin.conf and super-admin.conf to point to local API server
kubeadm init --config some-kubeadm-config.yaml # this will end up using the existing conf files
# sed back admin.conf and super-admin.conf to point to LB
```

**Comment by Macgyverops** [comment]
So does it work on 1.27?

**Comment by Cellebyte** [comment]
@Macgyverops the first control-plane of a cluster gets a super-admin.conf file in the host filesystem. The other control-planes only get the admin.conf as before. To circumvent the issue you need to make sure that on the first-control-plane kube-vip mounts the super-admin.conf instead of the normal admin.conf. After the cluster correctly initialized the first control-plane also receives the admin.conf and the static manifest can be reverted to use the normal admin.conf file.

**Comment by watjt** [comment]
@chrischdi Will there be any problems if I keep using the /etc/kubernetes/super-admin.conf file?

**Comment by chrischdi** [comment]
> [@chrischdi](https://github.com/chrischdi) Will there be any problems if I keep using the /etc/kubernetes/super-admin.conf file?

Yes and no, when joining a control-plane node, it simply won't have the file and if kube-vip runs there as static pod it won't work.

**Comment by geneliu** [comment]
Would like to flag this on. Even though the workaround (init with super-admin.yaml mount, and joint other control plane nodes with admin.yaml) works, is there on going investigation to make it work. Using cluster-api setup workload cluster with the work around mentioned above does not work as you need **manually** update the manifest.

**Comment by sathieu** [comment]
@geneliu You can take a look at capv :

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/v1.13.1/templates/cluster-template.yaml#L178-L224

(do not forget https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/v1.13.1/templates/cluster-template.yaml#L244-L245)

**Comment by Cellebyte** [comment]
@geneliu currently we are just tracking it.
For now our focus is not on this topic as we are working on refactoring, stability and maintenance possibility of the kube-vip code.

**Comment by geneliu** [comment]
> [@geneliu](https://github.com/geneliu) You can take a look at capv :
> 
> https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/v1.13.1/templates/cluster-template.yaml#L178-L224
> 
> (do not forget https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/v1.13.1/templates/cluster-template.yaml#L244-L245)

@sathieu These are very useful. Thank you!

**Comment by Whisper40** [comment]
Hello, any update for this to be fix ?

**Comment by briantopping** [comment]
This is a super annoying issue with an otherwise great open source product. The issue has been well-understood for almost two years. Is there something the community can do to help close this? Thanks!

**Comment by Cellebyte** [comment]
@briantopping as you already said, everything needed to start working on this problem is described in the issue. Additionally the last statements are still valid. It has not a high priority as a workaround exists. I understand it can be annoying, but it seems in the last 2 years no one tried to even fix it.

**Comment by briantopping** [comment]
Hmm. Yeah that’s tricky. I’m new to the product so it’s difficult for me to take a position. Maybe I shouldn’t have posted at all. What it seems like is it’s a very good product that has a very risky barrier to entry. I suspect a lot of people would try it and dismiss it as unusable before even realizing it was something simple that was holding them up. Some of them might be able to fix this if they became a part of the community. Or some other issue as they grew to depend on the product. Aren’t we all that way? 

It sounds like this has become a feature though – a kind of litmus test to see who will actually step up. And maybe the reason nobody has tried is because of the security implications of getting it wrong. I certainly don’t want to be responsible for a site breach because I blindly implemented a couple of `sed` scripts and pushed a PR. I’ve probably already spent more time talking about this than it might take for someone who knows the code and the requirements to fix. 

So you have my apologies for saying anything. I guess I’m left feeling the same way, that if nobody wants to fix this, maybe the project isn’t so good to depend on after all. I don’t mean that to insinuate anything, so please don’t take it personally. None of us “deserve” this to be fixed and I regret if my previous note sounded entitled. It just seems like a loss overall.

**Comment by thebsdbox** [comment]
👋 apologies, this is more my lack of making things complete. There was additional functionality added to kube-vip to improve this situation. It needs more documentation adding to make this a little clearer. I'll try and work on this during the week.

**Comment by Cellebyte** [comment]
@briantopping you were right pointing it out, but we’re slowly improving docs and config examples. And this is a good topic as well to be tackled after we now did the refactoring :)

**Comment by lentzi90** [comment]
This was indeed the first issue I hit when trying out kube-vip. I propose adding a note to the documentation as a first step (see https://github.com/kube-vip/website/pull/61). That way people will at least have an easier time figuring out what is wrong and finding this, than I did.

**Comment by ralgozino** [comment]
A tip that I did not see mentioned in this thread or in the docs: If you are using the CLI to generate the manifest, you can use the flag `--k8sConfigPath` to the set the path to the super-admin.conf file like so (following the example in the docs):

```bash
# add your export commands here setting the env vars as explained in the docs
# where `kube-vip` is the alias that the docs tell you to set up
kube-vip manifest pod \
        --interface ${INTERFACE} \
        --address ${VIP} \
        --controlplane \
        --services \
        --arp \
        --leaderElection \
        --k8sConfigPath=/etc/kubernetes/super-admin.conf | tee /etc/kubernetes/manifests/kube-vip.yaml
```

Just be sure to set the flag only for the first control-plane node (or set it to admin.conf path for the other two nodes).

After the init has been done, you can relaunch the command without the flag in the first control-plane to regenerate the manifest again but pointing to the admin.conf to keep the config aligned between nodes.

**Comment by HarishPampana** [comment]
The Root Problem
What Was Happening Before:
1) kube-vip was working correctly - it was binding the VIP (20.0.77.100) and acquiring leadership
2) kubeadm was configured incorrectly - you were using:
```
kubeadm init --control-plane-endpoint="testk8svip :6443"
```
3) DNS resolution timing issue - During bootstrap, kubeadm would:
         * Start the API server on the node's actual IP (20.0.77.101)
         * Try to check health at https://20.0.77.101:6443/livez
         * But then redirect to https://testk8svip :6443/livez due to the hostname configuration
         * This failed because kube-vip hadn't fully taken over the VIP yet during the bootstrap phase
            The Certificate Problem:
            The API server certificates were generated for:
        
      ```
          testk8svip (hostname)
          20.0.77.101 (node IP)
          But NOT for 20.0.77.100 (VIP IP)
       ```
So when kubeadm tried to connect via the VIP, the certificate validation failed.

How the Fix Worked
1) Added proper certificate SANs:
     #echo "20.0.77.100 testk8svip " >> /etc/hosts
     #echo "20.0.77.100 kubernetes" >> /etc/hosts
```
kubeadm init \
  --skip-phases=addon/kube-proxy \
  --control-plane-endpoint="20.0.77.100:6443" \
  --pod-network-cidr="192.168.0.0/16" \
  --apiserver-cert-extra-sans="testk8svip ,20.0.77.100,$(hostname)" \
  --v=2
```
Note: I skip the kube-proxy for CNI Cilium and It's optional.
